pkg/template/nodes.go

```diff
@@ -50,6 +59,22 @@ func (n *Nodes) FindNode(tag NodeTag) (EvaluationNode, bool) {
 	return node, ok
 }
 
+// AddAnnotation creates an entry in the annotations map of Nodes.
+// The entry is NodeAnnotation with only the annotation's position,
+// indexed by NodeTag and AnnotationName.
+func (n *Nodes) AddAnnotation(tag NodeTag, ann Annotation) {
+	if _, found := n.annotations[tag]; !found {
+		n.annotations[tag] = NodeAnnotations{}
+	}
+	n.annotations[tag][ann.Name] = NodeAnnotation{Position: ann.Position}
```

Comment:
im not sure i understand why we are using NodeAnnotation here. It seems like purpose of AddAnnotation is to store position mapped to a name of the annotation. shouldnt it be something like name => Position? so i would argue that it should be AddAnnotationPosition/FindAnnotationPosition.

additionally whats tripping me up is that annotation name is technically not a unique key, since user can specify multiple annotations on the same node -- should we be failing somewhere for that? may be we should be using index of the annotation for that node.

Reply:
> im not sure i understand why we are using NodeAnnotation here. It seems like purpose of AddAnnotation is to store position mapped to a name of the annotation. shouldnt it be something like name => Position? so i would argue that it should be AddAnnotationPosition/FindAnnotationPosition.

Your read is spot on.

We've been thinking of it slightly broader: that there are two parts of annotation data:
- the bits that are evaluated (args and kwargs)
- the bits that are not (position)

The stuff in that second category would... categorically... be carried over. So, even though it's one field, today, that felt more like an implementation detail rather than a conceptual distinction.

But yes, YAGNI — the only field being carried over is a position.

So, our current thinking is that the risk of accidentally depending on some new field in the source annotation is outweighed by the design conveying that it's annotation data making it easier to grok what's going on, here.

How does that show up on _your_ scorecard?


> additionally whats tripping me up is that annotation name is technically not a unique key, since user can specify multiple annotations on the same node -- should we be failing somewhere for that? may be we should be using index of the annotation for that node

We didn't want to introduce a breaking change with this work. Today, [the last annotation in, wins](https://carvel.dev/ytt/#gist:https://gist.github.com/pivotaljohn/5c5b8d0ed1bb47e5cd5204b54514f8cc).

This probably goes on the ytt v1 wish list.